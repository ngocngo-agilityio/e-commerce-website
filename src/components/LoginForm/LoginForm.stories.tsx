import type { Meta, StoryObj } from '@storybook/react';

import LoginForm from '.';

const meta = {
  title: 'LoginForm',
  component: LoginForm,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof LoginForm>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {},
};
