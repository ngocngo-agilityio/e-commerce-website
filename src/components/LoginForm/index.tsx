'use client';

import { memo, useCallback } from 'react';
import { Controller, useForm } from 'react-hook-form';
import { ViewIcon, ViewOffIcon } from '@chakra-ui/icons';
import {
  Flex,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Input,
  Text,
  Button,
  InputRightElement,
  InputGroup,
  useDisclosure,
} from '@chakra-ui/react';

// Constants
import { ERROR_MESSAGES, REGEX } from '@constants';

// Utils
import { isEnableSubmitButton } from '@utils';

export interface ISignInForm {
  email: string;
  password: string;
}

const REQUIRE_FIELDS = ['email', 'password'];

const SIGN_IN_VALIDATION_RULE = {
  email: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Email'),
    pattern: {
      value: REGEX.EMAIL,
      message: ERROR_MESSAGES.FIELD_INVALID('Email'),
    },
  },
  password: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Password'),
    minLength: { value: 8, message: ERROR_MESSAGES.PASSWORD_NOT_LONG },
  },
};

export interface ISignInFormProps {
  isSubmitting?: boolean;
  onSignIn: (data: ISignInForm) => void;
}

const LoginForm = ({
  isSubmitting = false,
  onSignIn,
}: ISignInFormProps): JSX.Element => {
  const { isOpen: isShowPassword, onToggle: onTogglePassword } =
    useDisclosure();

  const {
    handleSubmit,
    control,
    clearErrors,
    formState: { dirtyFields, errors },
  } = useForm<ISignInForm>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    defaultValues: {
      email: '',
      password: '',
    },
  });

  const dirtyItems = Object.keys(dirtyFields).filter(
    (key) => dirtyFields[key as keyof ISignInForm],
  );
  const shouldEnable = isEnableSubmitButton(REQUIRE_FIELDS, dirtyItems, errors);
  const isDisableSubmit = !shouldEnable || isSubmitting;

  // Clear error when typing that field.
  const handleOnChange = useCallback(
    (fieldName: keyof ISignInForm) => {
      clearErrors(fieldName);
    },
    [clearErrors],
  );

  const handleSignIn = useCallback(
    (data: ISignInForm) => onSignIn(data),
    [onSignIn],
  );

  return (
    <Flex
      as="form"
      flexDir="column"
      w={{ base: 'full', md: '460px' }}
      gap={7}
      onSubmit={handleSubmit(handleSignIn)}
    >
      <Flex flexDir="column">
        <FormLabel>Email</FormLabel>

        <Controller
          name="email"
          control={control}
          rules={SIGN_IN_VALIDATION_RULE.email}
          render={({ field: { onChange, ...rest }, fieldState: { error } }) => (
            <FormControl isInvalid={!!error}>
              <Input
                {...rest}
                placeholder="Your email"
                isDisabled={isSubmitting}
                onChange={(e) => {
                  const value = e.target?.value;

                  handleOnChange('email');
                  onChange(value);
                }}
              />

              <FormErrorMessage>{error?.message}</FormErrorMessage>
            </FormControl>
          )}
        />
      </Flex>

      <Flex flexDir="column">
        <FormLabel>Password</FormLabel>
        <Controller
          name="password"
          control={control}
          rules={SIGN_IN_VALIDATION_RULE.password}
          render={({ field: { onChange, ...rest }, fieldState: { error } }) => (
            <FormControl isInvalid={!!error}>
              <InputGroup>
                <Input
                  {...rest}
                  type={isShowPassword ? 'text' : 'password'}
                  placeholder="Enter password"
                  isDisabled={isSubmitting}
                  onChange={(e) => {
                    const value = e.target?.value;

                    handleOnChange('password');
                    onChange(value);
                  }}
                />
                <InputRightElement
                  onClick={onTogglePassword}
                  style={{ cursor: 'pointer' }}
                  data-testid="view-icon"
                >
                  {isShowPassword ? (
                    <ViewOffIcon boxSize={5} />
                  ) : (
                    <ViewIcon boxSize={5} />
                  )}
                </InputRightElement>
              </InputGroup>

              <FormErrorMessage>{error?.message}</FormErrorMessage>
            </FormControl>
          )}
        />
      </Flex>

      <Text
        pt={3}
        textAlign="end"
        color="black"
        fontWeight="bold"
        fontSize="md"
      >
        Forgot password
      </Text>

      <Button
        type="submit"
        borderRadius="sm"
        w="full"
        fontSize="xl"
        isDisabled={isDisableSubmit}
        isLoading={isSubmitting}
      >
        Sign in
      </Button>
    </Flex>
  );
};

export default memo(LoginForm);
