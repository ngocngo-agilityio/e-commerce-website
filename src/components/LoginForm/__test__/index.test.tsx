import {
  render,
  fireEvent,
  screen,
  waitFor,
  act,
} from '@testing-library/react';

// Mocks
import { MOCK_SIGN_IN_PAYLOAD } from '@mocks';

// Constants
import { ERROR_MESSAGES } from '@constants';

// Components
import LoginForm from '../index';

describe('LoginForm', () => {
  const mockOnSignIn = jest.fn();

  const submitForm = (email: string, password: string) => {
    const EMAIL_FIELD = screen.getByPlaceholderText('Your email');
    const PASSWORD_FIELD = screen.getByPlaceholderText('Enter password');
    const TOGGLE_PASSWORD_ICON = screen.getByTestId('view-icon');
    const SUBMIT_BTN = screen.getByText('Sign in');

    act(() => {
      // Change email
      fireEvent.change(EMAIL_FIELD, {
        target: { value: email },
      });
      fireEvent.blur(EMAIL_FIELD);

      // Change password
      fireEvent.change(PASSWORD_FIELD, {
        target: { value: password },
      });
      fireEvent.blur(PASSWORD_FIELD);

      fireEvent.click(TOGGLE_PASSWORD_ICON);

      fireEvent.click(SUBMIT_BTN);
    });
  };

  test('should match with snapshot', () => {
    const { container } = render(<LoginForm onSignIn={mockOnSignIn} />);

    expect(container).toMatchSnapshot();
  });

  test('should call onSignIn when submit the valid form', () => {
    render(<LoginForm onSignIn={mockOnSignIn} />);

    // Change form
    submitForm(MOCK_SIGN_IN_PAYLOAD.email, MOCK_SIGN_IN_PAYLOAD.password);

    waitFor(() => {
      expect(mockOnSignIn).toHaveBeenCalledWith(MOCK_SIGN_IN_PAYLOAD);
    });
  });

  test('should show errors when blur with the invalid field', async () => {
    render(<LoginForm onSignIn={mockOnSignIn} />);

    // Change form
    submitForm('a', 'a');

    await waitFor(() => {
      expect(ERROR_MESSAGES.FIELD_INVALID('Email')).toBeTruthy();
      expect(ERROR_MESSAGES.PASSWORD_NOT_LONG).toBeTruthy();
    });
  });
});
