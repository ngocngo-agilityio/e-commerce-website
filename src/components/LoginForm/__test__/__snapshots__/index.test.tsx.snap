// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match with snapshot 1`] = `
<div>
  <form
    class="css-1m7lw4y"
  >
    <div
      class="css-j7qwjs"
    >
      <label
        class="chakra-form__label css-uvho5g"
      >
        Email
      </label>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <input
          class="chakra-input css-0"
          id="field-:r1:"
          name="email"
          placeholder="Your email"
          value=""
        />
      </div>
    </div>
    <div
      class="css-j7qwjs"
    >
      <label
        class="chakra-form__label css-uvho5g"
      >
        Password
      </label>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            class="chakra-input css-0"
            id="field-:r2:"
            name="password"
            placeholder="Enter password"
            type="password"
            value=""
          />
          <div
            class="chakra-input__right-element css-dp9r5f"
            data-testid="view-icon"
            style="cursor: pointer;"
          >
            <svg
              class="chakra-icon css-w31a5m"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <g
                fill="currentColor"
              >
                <path
                  d="M23.432,10.524C20.787,7.614,16.4,4.538,12,4.6,7.6,4.537,3.213,7.615.568,10.524a2.211,2.211,0,0,0,0,2.948C3.182,16.351,7.507,19.4,11.839,19.4h.308c4.347,0,8.671-3.049,11.288-5.929A2.21,2.21,0,0,0,23.432,10.524ZM7.4,12A4.6,4.6,0,1,1,12,16.6,4.6,4.6,0,0,1,7.4,12Z"
                />
                <circle
                  cx="12"
                  cy="12"
                  r="2"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <p
      class="chakra-text css-s0uli3"
    >
      Forgot password
    </p>
    <button
      class="chakra-button css-g5yz1r"
      disabled=""
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;
